src/main/java/org/codeforamerica/messaging/services/EmailService.java

```diff
@@ -41,4 +50,26 @@ private boolean unsubscribed(String toEmail) {
         EmailSubscription latestSubscription = emailSubscriptionRepository.findFirstByEmailOrderByCreationTimestampDesc(toEmail);
         return latestSubscription != null && latestSubscription.isUnsubscribed();
     }
+
+    @Transactional
+    public void updateStatus(String providerMessageId, MessageStatus newEmailStatus, String rawEmailStatus, Map<String, String> providerError) {
+        EmailMessage emailMessage = emailMessageRepository.findFirstByProviderMessageId(providerMessageId);
+        if (emailMessage == null) {
+            log.error("Cannot find message with providerId: {}", providerMessageId);
+            throw new JobRunrException("Cannot find message with providerId");
+        }
+        log.info("Updating status for {}", providerMessageId);
+        MessageStatus currentEmailStatus = emailMessage.getMessage().getEmailStatus();
+        if (newEmailStatus.isAfter(currentEmailStatus)) {
+            log.info("Updating status. Provider message id: {}, current status: {}, new status: {}", providerMessageId, currentEmailStatus, newEmailStatus);
+            emailMessage.getMessage().setRawEmailStatus(rawEmailStatus);
+            emailMessage.getMessage().setEmailStatus(newEmailStatus);
+            emailMessage.setProviderError(providerError);
+            emailMessageRepository.save(emailMessage);
```

Comment:
Is the version auto managed here? Only updated if a change is made? And the `@Transactional` tag means that if the version changed while this was running, then this job will fail and restart with the new version of the record?

Reply:
Yes, and yes.